jgrapht-core/src/main/java/org/jgrapht/alg/flow/PushRelabelMaximumFlow.java

```diff
@@ -0,0 +1,353 @@
+package org.jgrapht.alg.flow;
+
+import org.jgrapht.DirectedGraph;
+import org.jgrapht.alg.util.Extension.ExtensionFactory;
+import org.jgrapht.alg.util.Pair;
+
+import java.util.*;
+
+/**
+ * <p><a href="https://en.wikipedia.org/wiki/Push%E2%80%93relabel_maximum_flow_algorithm">
+ * Push-relabel maximum flow</a> algorithm designed by Andrew V. Goldberg and
+ * Robert Tarjan. Current implementation complexity upper-bound is O(V^3). For more
+ * details see: <i>"A new approach to the maximum flow problem"</i> by Andrew V. Goldberg
+ * and Robert Tarjan <i>STOC '86: Proceedings of the eighteenth annual ACM symposium on
+ * Theory of computing</i></p>
+ *
+ */
+public class PushRelabelMaximumFlow<V, E> extends MaximumFlowAlgorithmBase<V,E> {
+
+    private static final boolean LABEL_PRUNE_ENABLED = true;
+
+    private DirectedGraph<V, E> network;
+
+    private final ExtensionFactory<VertexExtension> vertexExtensionsFactory;
+    private final ExtensionFactory<EdgeExtension>   edgeExtensionsFactory;
+
+    // Label pruning
+
+    private Map<Integer, Integer> labeling;
+
+    boolean flowBack;
+
+    // Diagnostic
+
+    private static final boolean DIAGNOSTIC_ENABLED = false;
+
+    private class PushRelabelDiagnostic {
+
+        // Discharges
+        Map<Pair<V, V>, Integer> discharges = new HashMap<Pair<V, V>, Integer>();
```

Comment:
There already exists a class called VertexPair in Jgrapht. Could you use that one here instead?
I noticed that you also propose to include a general class 'Pair'. Perhaps it makes sense to let the VertexPair class extend your Pair class?